### PR TITLE
feat(ai-fix): add AI-fix patch wire-format schema

### PR DESCRIPTION
**Pull request by Jayclifford345**
## What & why

**PR3 of the PR-886 AI auto-heal breakdown** ([RFC](https://github.com/grafana/pathfinder-rfcs/blob/208096ae619519c32ff2f56aca84685bd4a2414d/plans/PR-886-AI-AUTOHEAL-BREAKDOWN.md)). PR1 ([#980](https://github.com/grafana/grafana-pathfinder-app/pull/980)) is merged. PR2 (id-space reconciliation) was **folded into PR4** — `main` already settles the id space (`deriveStepId` + id-keyed completion + `reconcileSection` orphan self-heal), so the canonical-id/block-finder work moves into PR4 where `applyPatchToGuide` lives.

This PR adds the **AI-fix patch wire-format schema** — the Zod contract for what the Grafana Assistant is allowed to return when a user clicks the AI "Fix this" button on a failing interactive step.

## What's included

Two net-new files under `src/integrations/assistant-integration/`:
- `ai-fix-patch.schema.ts` — a `discriminatedUnion('type')` of:
  - `selector-patch` — replace a failing step's `reftarget`
  - `substep-selector-patch` — replace a `reftarget` inside a multistep/guided container (addressed by container id + positional index)
  - `prepend-step` — insert a new interactive step before the failing one; `newStep` round-trips through the authoritative `JsonInteractiveBlockSchema`
- `ai-fix-patch.schema.test.ts` — boundary tests

Every selector field passes `SafeSelectorSchema`: a deny-list (`<`, `>`, `` ` ``, `${`, `javascript:`, `data:`, `vbscript:`) + a 512-char cap. This is the security core — these strings flow into `querySelector`, so anything HTML/URL/template-shaped is rejected at the boundary.

## It lands dark

Net-new and inert — nothing imports `AiFixPatchSchema` until PR4 (`applyPatchToGuide`) and PR5 (the generation hook). No barrel export added yet (those PRs import the module directly). Zero behavior change for existing users.

## Test plan

- `npm run typecheck` ✅ — confirms `prepend-step`'s `JsonInteractiveBlockSchema.superRefine(...)` compiles against `main`'s Zod-4 schema
- `npm run lint` ✅ 0 errors · `prettier` ✅
- `jest ai-fix-patch.schema.test.ts` ✅ 12/12 — deny-list rejects HTML / `javascript:` / `${…}` / >512-char; `prepend-step` delegates to `JsonInteractiveBlockSchema` (non-`noop`-without-`reftarget` rejects, `noop`-without-`reftarget` accepts); unknown discriminator rejects
- governance/architecture (Tier 3 → Tier 0 import) ✅ 73/73

### CODE REVIEW

File: src/integrations/assistant-integration/ai-fix-patch.schema.test.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        type: 'selector-patch',
+        targetStepId: 'step-1',
+        newReftarget: '.x'.repeat(300),
+      });
+      expect(result.success).toBe(false);
+    });
+  });
+
+  describe('substep-selector-patch', () => {
+    it('accepts a well-formed substep selector patch', () => {
```
Comment (moxious) [inline src/integrations/assistant-integration/ai-fix-patch.schema.test.ts:70]: add tests: invalid values for subStepIndex = (-1, null) should fail; that containerId = '' or containerId = null should fail.

MIght consider to prove that SafeParse gets called on the selector in the subStep example, but that's mostly pretty safe because of how it's reused.

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      }
+    }
+  }),
+  rationale: z.string().max(500).optional(),
+});
+
+const SubstepSelectorPatchSchema = z.object({
+  type: z.literal('substep-selector-patch'),
+  containerId: z.string().min(1),
+  subStepIndex: z.number().int().min(0),
```
Comment (moxious) [inline src/integrations/assistant-integration/ai-fix-patch.schema.ts:53]: maybe add comment for later PR:   TODO you have to validate that this is in bounds and not let people pass subStepIndex=100000

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import { JsonInteractiveBlockSchema } from '../../types/json-guide.schema';
+
+// Deny-list: reject HTML, template interpolation, and unsafe URL schemes (XSS guard); 512-char cap bounds querySelector cost.
+const DANGEROUS_SELECTOR_PATTERNS = ['<', '>', '`', '${', 'javascript:', 'data:', 'vbscript:'] as const;
+
+const SafeSelectorSchema = z
+  .string()
+  .min(1)
+  .max(512, 'Selector exceeds 512-character cap')
+  .refine((sel) => !DANGEROUS_SELECTOR_PATTERNS.some((p) => sel.toLowerCase().includes(p)), {
```
Comment (moxious) [inline src/integrations/assistant-integration/ai-fix-patch.schema.ts]: `includes` might be a little too blunt force here.   

This is a valid selector that's rejected:  `a[href^="javascript:"]`

I see what you're trying to do security wise here, you might discriminate between dangerous selectors which start with a certain string (javascript, data) vs. those that contain it.

**Review by moxious — CHANGES_REQUESTED** [review]
very close to ready to merge imo; some very small suggestions on improving test coverage and a boundary condition before LGTM

File: src/integrations/assistant-integration/ai-fix-patch.schema.test.ts (hunk 1 above)
Comment (Jayclifford345) [reply src/integrations/assistant-integration/ai-fix-patch.schema.test.ts:70]: Added in 262f1ef8: `subStepIndex` = -1 and `null` now reject, `containerId` = `` and `null` reject, plus a test that a dangerous `newReftarget` inside a substep patch is rejected (proves `SafeSelectorSchema` runs on substeps too, not just top-level patches).

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts (hunk 2 above)
Comment (Jayclifford345) [reply src/integrations/assistant-integration/ai-fix-patch.schema.ts:53]: The upper bound can't be checked at the schema layer — it has no view of the container's substep count. It's enforced one PR up in #984's `applyPatchToGuide`: `apply-ai-fix-patch.ts` returns an `out-of-range` error when `subStepIndex >= block.steps.length`, covered by the test "fails with a range error when subStepIndex is out of bounds" in `apply-ai-fix-patch.test.ts`. Skipped the TODO comment to stay within the repo's no-comment policy since the behavior already lands.

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts (hunk 3 above)
Comment (Jayclifford345) [reply src/integrations/assistant-integration/ai-fix-patch.schema.ts]: Good catch. Split the deny-list in 262f1ef8: `javascript:`/`data:`/`vbscript:` are now matched as a *prefix* (after trim + lowercase), so `a[href^="javascript:"]` passes while `javascript:alert(1)` is still rejected. HTML/template chars (`<`, `>`, backtick, `${`) stay blocked anywhere. Added a test locking in the attribute-selector case.

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      return !startsWithUnsafeScheme && !containsUnsafeSubstring;
+    },
+    { error: 'Selector contains a disallowed substring (HTML, template, or unsafe URL scheme)' }
+  );
+
+const SelectorPatchSchema = z.object({
+  type: z.literal('selector-patch'),
+  targetStepId: z.string().min(1),
+  newReftarget: SafeSelectorSchema,
+  rationale: z.string().max(500).optional(),
```
Comment (moxious) [inline src/integrations/assistant-integration/ai-fix-patch.schema.ts:29]: chatty LLM losing the whole fix because rationale ran one char long could be a problem.  Might want to allow anything and then use `.transform(s => s.slice(0, 500))` to cap effectively what will be seen instead of failing schema validation.   If an LLM gets too verbose, what behavior do you want?

**Review by moxious — APPROVED** [review]
I had one minor comment which shouldn't block because it's so minor, but worth taking a look at to confirm intent.

File: src/integrations/assistant-integration/ai-fix-patch.schema.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import { z } from 'zod';
+
+import { JsonInteractiveBlockSchema } from '../../types/json-guide.schema';
+
+// XSS guard for strings flowing into querySelector: HTML/template chars are blocked
+// anywhere; URL schemes are blocked only as a prefix, so attribute matches like
+// a[href^="javascript:"] pass. 512-char cap bounds querySelector cost.
+const DANGEROUS_URL_SCHEMES = ['javascript:', 'data:', 'vbscript:'] as const;
```
Comment (leslie-heinzen) [inline src/integrations/assistant-integration/ai-fix-patch.schema.ts:8]: Kinda surprised the grafana CSP doesn't catch these. Does it allow unsafe-inline for some reason? 👀  

Looks fine, just curious.

Also, vbscript: 

Image
